### PR TITLE
[Pooling] Improve BGE-M3 sync pooling throughput by up to 3.13%

### PR DESCRIPTION
**Pull request by taneem-ibrahim**
## Purpose

Improve BGE-M3 and other CPU-token-ID pooling workloads on pooling Model Runner V2 by avoiding unnecessary model-device prompt-token-ID metadata.

Today `requires_token_ids=True` makes MRV2 preserve prompt token IDs for poolers that need token-level postprocessing. BGE-M3's sparse embedding path uses SPLADE, which only needs prompt token IDs on CPU, so copying the full prompt-token-ID batch to GPU adds unnecessary serving overhead.

For Jina/BGE-M3 ranking, doc/query marker indexes are computed from CPU prompt token IDs, and only the final concatenated index tensor is copied to GPU before indexing hidden states. This keeps the implementation aligned with MRV2's no-consistent-batch design while preserving the existing pooling behavior.

### Accuracy

```bash
CUDA_VISIBLE_DEVICES=1 VLLM_USE_V2_MODEL_RUNNER=1 \
  .venv/bin/python -m pytest \
  tests/models/transformers/test_backend.py::test_pooling -v
```

```text
================== 4 passed, 15 warnings in 139.13s (0:02:19) ==================
```

Ran a dedicated numerical `main` vs branch BGE-M3 accuracy diff with MRV2 enabled.

- Prompt count: `8`
- Prompt SHA256: `fd8c88bf564e8d7261d16beecbebf91233ce02c45ad3734eaaadebbf7d570f01`
- Dense/sparse max-abs tolerance: `5e-4`
- Result: `PASS`

| Output | Compared | Max abs diff | Max relative diff | Result |
|---|---:|---:|---:|---|
| Standard dense embeddings | 8 vectors / 8192 values | 0.000149 | 1.42 | PASS |
| BGE-M3 plugin dense embeddings | 8 vectors / 8192 values | 0.000244 | 1.42 | PASS |
| BGE-M3 plugin sparse embeddings | 8 rows / 110 weights | 0.000488 | 0.0329 | PASS |

Sparse output details:

| Field | Result |
|---|---|
| Sparse token sets | Matched |
| Sparse token text | Matched |
| Nonzero range per row | `[3, 28]` |

The accuracy results are numerically stable within `5e-4` max-abs tolerance

### Perf: Performance Comparisons Results

Each benchmark request contains 32 RuBQ documents. Documents/second is `request_throughput * 32`. Each row below is the median of 3 repetitions, and each repetition completed all 128 requests with 0 failures.

Benchmark client:

```bash
bench_bge() {
  local label="$1"
  local concurrency="$2"

  for repetition in 1 2 3; do
    .venv/bin/vllm bench serve \
      --backend openai-embeddings \
      --base-url "http://127.0.0.1:${PORT}" \
      --endpoint /v1/embeddings \
      --model bge \
      --tokenizer BAAI/bge-m3 \
      --dataset-name custom \
      --dataset-path /tmp/bge-bench/rubq_batches32.jsonl \
      --disable-shuffle \
      --skip-chat-template \
      --num-prompts 128 \
      --num-warmups 4 \
      --request-rate inf \
      --max-concurrency "$concurrency" \
      --percentile-metrics e2el \
      --metric-percentiles 50,99 \
      --disable-tqdm \
      --save-result \
      --result-dir /tmp/bge-bench/results \
      --result-filename "${label}_rep${repetition}.json"
  done
}
```

MRV2 sync server, run once on `origin/main` and once on this branch:

```bash
CUDA_VISIBLE_DEVICES="$GPU" \
VLLM_USE_V2_MODEL_RUNNER=1 \
.venv/bin/vllm serve BAAI/bge-m3 \
  --served-model-name bge \
  --runner pooling \
  --host 127.0.0.1 \
  --port "$PORT" \
  --api-server-count 8 \
  --no-async-scheduling
```
Benchmarked current `main` against this branch on the same NVIDIA H200 GPU, same RuBQ dataset, same server/client commands, and 3 repetitions per configuration.

- Dataset: `mteb/RuBQRetrieval`, first 4096 corpus documents
- Request shape: 128 requests, 32 documents/request
- Metric: documents/second = `request_throughput * 32`
- Main commit: `39eba6ac368133ab4e69a0589fcd53c481dddc44`
- Branch commit: `71b2ce24429a0b821cef7ae55aacb64e0429630d`

| Mode | Concurrency | Main median docs/s | Branch median docs/s | Delta | Main median p50 | Branch median p50 | Main median p99 | Branch median p99 |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| MRV2 sync | 16 | 3229.0 | 3330.1 | +3.13% | 144.2 ms | 129.0 ms | 231.2 ms | 264.1 ms |
| MRV2 sync | 32 | 3222.0 | 3249.7 | +0.86% | 264.2 ms | 263.8 ms | 568.8 ms | 633.8 ms |

The primary pooling comparison is the sync server configuration, where this branch improves median throughput by +3.13% at concurrency 16 and ~1% at concurrency 32. 

Per-repetition docs/s:

| Configuration | Main docs/s | Branch docs/s |
|---|---:|---:|
| MRV2 sync c16 | `[2992.6, 3229.0, 3350.4]` | `[3033.5, 3330.1, 3355.0]` |
| MRV2 sync c32 | `[3222.0, 3169.0, 3283.7]` | `[3375.2, 3213.6, 3249.7]` |

Deltas:
sync c16: +3.13%
sync c32: ~1%

## AI assistance disclosure

OpenAI Codex (GPT-5) assisted with finding optimization exploration pathways and drafting the code,. I reviewed every changed line and ran all tests and benchmarks listed above.

### CODE REVIEW

File: vllm/pooling_params.py [hunk 1]
```diff
@@ -65,6 +65,7 @@ class PoolingParams(
     ## Internal use only
     task: PoolingTask | None = None
     requires_token_ids: bool = False
+    requires_token_ids_gpu: bool = False
```
Comment (yewentao256) [inline vllm/pooling_params.py]: Is there any way we can remove this attribute? One design change from v1 to v2 is that we remove the consistent batch.

Comment (yewentao256) [reply vllm/pooling_params.py]: Maybe we can calculate `query_indexes` in cpu and h2d the `indexes` only

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!

File: vllm/pooling_params.py (hunk 1 above)
Comment (taneem-ibrahim) [reply vllm/pooling_params.py]: Great advice. I removed the MRV2 GPU token-id requirement path by keeping required prompt token IDs on CPU, computing doc/query indexes on CPU, and then async H2D copying only the final selected index tensor.

**Review by yewentao256 — COMMENTED** [review]
Nice! Shall we delete `PoolingMetadata.prompt_token_ids` and `get_prompt_token_ids` now?

**Comment by DarkLight1337** [comment]
PTAL @noooop

File: vllm/v1/pool/metadata.py [hunk 2]
```diff
@@ -17,6 +17,8 @@ class PoolingCursor:
     prompt_lens_cpu: torch.Tensor
     seq_lens_cpu: torch.Tensor
     num_scheduled_tokens_cpu: torch.Tensor
+    partial_prefill: bool | None = None
+    finished_mask: list[bool] | None = None
```
Comment (noooop) [inline vllm/v1/pool/metadata.py:22]: Thank you for eliminating the use of prompt_token_ids(gpu). 

When would partial_prefill and finished_mask be None? 

Perhaps disallowing these two variables from being None could simplify the comparison.

File: vllm/v1/pool/metadata.py [hunk 3]
```diff
@@ -130,9 +141,14 @@ def build_pooling_cursor(
...
+        seq_lens_np = seq_lens_cpu.numpy()
         if query_start_loc_gpu is None:
             cumsum = torch.zeros(
-                n_seq + 1, dtype=torch.int64, pin_memory=PIN_MEMORY, device="cpu"
+                n_seq + 1,
+                dtype=torch.int64,
+                pin_memory=PIN_MEMORY and device.type != "cpu",
+                device="cpu",
             )
             torch.cumsum(num_scheduled_tokens_cpu, dim=0, out=cumsum[1:])
```
Comment (noooop) [inline vllm/v1/pool/metadata.py:153]: Why do we need to change this?

File: vllm/v1/pool/metadata.py (hunk 2 above)
Comment (taneem-ibrahim) [reply vllm/v1/pool/metadata.py:22]: Thanks, Yuqi. This is a good idea. I tightened this up so `PoolingCursor` always has concrete `partial_prefill` and `finished_mask` values, and slices just recompute those tiny CPU-side checks instead of using `None` as an “unknown” state.

File: vllm/v1/pool/metadata.py (hunk 3 above)
Comment (taneem-ibrahim) [reply vllm/v1/pool/metadata.py:153]: I changed it while making the pooling cursor path friendlier to CPU-only unit tests: when the target device is CPU, pinned memory does not help and can fail in environments without CUDA support. For GPU serving the behavior is unchanged because `device.type != "cpu"` still enables pinned memory.

I will revert it since it won't make any difference for the performance itself.

**Review by noooop — APPROVED** [review]

File: vllm/v1/pool/metadata.py (hunk 3 above)
Comment (noooop) [reply vllm/v1/pool/metadata.py:153]: After searching the vLLM repo and finding that no such patch is applied, I think this is unnecessary.
